File: 1448-CountGoodNodesinBinaryTree/1448-CountGoodNodesinBinaryTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def goodNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def dfs(root, maxSeenSoFar):
            count = 0
            if root.val >= maxSeenSoFar:
                count += 1
                maxSeenSoFar = max(maxSeenSoFar, root.val)

            if root.left:
                count += dfs(root.left, maxSeenSoFar)
            if root.right:
                count += dfs(root.right, maxSeenSoFar)
            
            return count
        
        return dfs(root, -float("inf"))

File: 1448-CountGoodNodesinBinaryTree/1448-CountGoodNodesinBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int goodNodes(TreeNode root) {
        return goodNodesHelper(root, -10001);
    }

    public int goodNodesHelper(TreeNode root, int max) {
        if (root == null) {
            return 0;
        }

        int count = 0;
        max = Math.max(max, root.val);
        if (root.val >= max) {
            count = 1;
        }

        return count + goodNodesHelper(root.left, max) + goodNodesHelper(root.right, max);
    }
}